File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/RespiratoryRateRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.RespiratoryRateRecord
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.mappers.toNumericDto
import com.phamtunglam.health_connector_hc_android.pigeon.RespiratoryRateRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [RespiratoryRateRecord] to a [RespiratoryRateRecordDto].
 */
internal fun RespiratoryRateRecord.toDto(): RespiratoryRateRecordDto = RespiratoryRateRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    rate = rate.toNumericDto(),
)

/**
 * Converts a [RespiratoryRateRecordDto] to a Health Connect [RespiratoryRateRecord].
 */
internal fun RespiratoryRateRecordDto.toHealthConnect(): RespiratoryRateRecord =
    RespiratoryRateRecord(
        rate = rate.value,
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(),
    )

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/respiratory_rate_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_platform_api.g.dart';

/// Mapper extensions for [RespiratoryRateRecord].
extension RespiratoryRateRecordMapper on RespiratoryRateRecord {
  /// Converts [RespiratoryRateRecord] to its [RespiratoryRateRecordDto].
  RespiratoryRateRecordDto toDto() {
    return RespiratoryRateRecordDto(
      id: id.toDto(),
      time: time.toUtc().millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      rate: rate.toDto() as NumericDto,
    );
  }
}

/// Mapper extensions for [RespiratoryRateRecordDto].
extension RespiratoryRateRecordDtoMapper on RespiratoryRateRecordDto {
  /// Converts [RespiratoryRateRecordDto] to its domain representation.
  RespiratoryRateRecord toDomain() {
    return RespiratoryRateRecord(
      id: HealthRecordId(id ?? HealthRecordId.none.value),
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true).toLocal(),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      rate: RespiratoryRate(breathsPerMinute: rate.value),
    );
  }
}
